### PR TITLE
add profiling_utils.configure and on_start_step

### PR DESCRIPTION
**Pull request by eundersander**
## Motivation and Context

We want to control the range of train steps that we capture for profiling.

Example usage for capturing a certain range of train steps (from step 200 to
step 300):
```
@profiling_utils.RangeContext("train")
def train():
    profiling_utils.configure(capture_start_step=200, num_steps_to_capture=100)
    initialize_training()
    for _ in range(10000)):
        profiling_utils.on_start_step()
        with profiling_utils.RangeContext("collect_rollout_steps"):
            collect_rollout_steps()
        with profiling_utils.RangeContext("update_agent"):
            update_agent()
```

## How Has This Been Tested

Added a unit test. Also some ad hoc testing with DDPPO training.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Hab-sim code freeze lifted

### CODE REVIEW

File: habitat_sim/utils/profiling_utils.py [hunk 1]
```diff
@@ -30,10 +30,34 @@ def my_function():
...
+    for _ in range(10000)):
+        profiling_utils.on_start_step()
+        with profiling_utils.RangeContext("collect_rollout_steps"):
+            collect_rollout_steps()
+        with profiling_utils.RangeContext("update_agent"):
+            update_agent()
+
+Example of capturing an Nsight Systems profile using your capture range:
+export HABITAT_PROFILING=1
+NSYS_NVTX_PROFILER_REGISTER_ONLY=0  # required when using capture range
```
Comment (eundersander) [inline habitat_sim/utils/profiling_utils.py:50]: todo: fix missing `export` here

File: habitat_sim/utils/profiling_utils.py [hunk 2]
```diff
@@ -44,21 +68,96 @@ def my_function():
...
+
+    If capture_start_step == 0, then capturing starts at the start of step 0
+    (after initialization code). If capture_start_step == -1, then capturing
+    starts immediately after your call to configure() and includes any
+    initialization code that runs prior to step 0. If num_steps_to_capture ==
+    -1, the range won't be popped and capture will continue until your program
+    terminates.
+    """
+
+    if not enable_profiling:
```
Comment (jturner65) [inline habitat_sim/utils/profiling_utils.py:100]: Is "enable_profiling"  a constant restricted to this module?  If so, would it not be better to have it named 
`_ENABLE_PROFILING` 

as per PEP? (with single leading underscore acting sort of like protected access mod - not sure if this variable is to be imported when this module is imported)

File: tests/test_profiling_utils.py [hunk 3]
```diff
@@ -30,33 +33,33 @@ def test_env_var_enable():
...
-    # Unfortunately, there isn't a way to verify correct behavior here. Ranges
-    # get recorded in the Nvidia driver/profiler. Documentation for
-    # torch.cuda.nvtx.range_push claims that it returns range stack depth, so I
-    # hoped to use it as a behavior checker, but it seems to just return -1 or
-    # -2 in practice.
 
     os.environ[env_var_name] = "1"
     importlib.reload(profiling_utils)
 
+    assert profiling_utils._helper.range_depth == 0
```
Comment (jturner65) [inline tests/test_profiling_utils.py:54]: Here too, isn't "env_var_name" a module level constant?

**Review by jturner65 — APPROVED** [review]
Just left a few comments regarding some constants. Otherwise, LGTM.

**Review by aclegg3 — APPROVED** [review]
👍

File: habitat_sim/utils/profiling_utils.py (hunk 2 above)
Comment (eundersander) [reply habitat_sim/utils/profiling_utils.py:100]: I wouldn't consider this a constant based on my usage: the value is determined at runtime and varies from run to run. But I agree about it being "internal use" so I'll add the underscore.

**Comment by eundersander** [comment]
I re-ran the failed CI "Run sim benchmark" test locally and it seems fine. I suspect this CI test is flaky.

**Pull request merged**